File: packages/kits/NeuronKit/Tests/NeuronKitTests/DreamingAlphaGateTests.swift
// DreamingAlphaGateTests.swift — T9 REM-ALPHA pending-count gate tests
// The ALPHA gate requires both a due timer and a non-empty dreaming queue.
//
// The §12.2 gate skips the dreaming cycle when the dreaming queue is empty
// (or not yet mounted), EVEN when the timer interval is due. Only a non-empty
// queue proceeds to the full dreaming-daemon pump + drain path.
//
// Coverage:
//   §1  POSITIVE: dreaming queue has pending items AND timer due
//                 → governor tick FIRES the dreaming cycle (dreamingFired=true).
//   §2  NEGATIVE: dreaming queue EMPTY AND timer due
//                 → governor tick does NOT fire (dreamingFired=false).
//                 Anti-regression: this test MUST fail if the gate is removed.
//   §3  CADENCE:  dreaming queue has pending items BUT timer interval not
//                 elapsed → governor tick does NOT fire (dreamingFired=false).

import Foundation
import Testing
import GeniusLocusKit
import LocusKit
import PersistenceKit
import PersistenceKitInMemory
@testable import NeuronKit

@Suite("T9 REM-ALPHA Pending-Count Gate — AutonomicGovernor §12.2")
struct DreamingAlphaGateTests {

    // MARK: - Infrastructure

    /// Open a fresh in-memory estate and return (kit, handle).
    private func makeKit() async throws -> (GeniusLocusKit, EstateHandle) {
        let kit = GeniusLocusKit()
        let owner = OwnerCredentials(ownerIdentifier: "alpha-gate-test")
        let config = EstateConfiguration(estateID: UUID(), backend: .inMemory)
        let storage = InMemoryStorage(configuration: config)
        _ = try await LocusKit.Estate.create(storage: storage, owner: owner)
        let handle = try await kit.open(storage: storage, owner: owner)
        return (kit, handle)
    }

    /// Capture a drawer into the estate.
    private func capture(
        _ kit: GeniusLocusKit,
        _ handle: EstateHandle,
        content: String
    ) async throws {
        let frame = CaptureFrame(
            content: content,
            channel: .typed,
            room: "test-room",
            latticeAnchor: .udc("000"),
            addedBy: "alpha-gate-test",
            embeddingModelID: "test-model-v1"
        )
        _ = try await kit.capture(handle, frame)
    }

    /// Fire one external-origin recall, enqueueing a DreamingItem for the
    /// co-surfaced drawer set (B-10a: enqueue fires ONLY on external origin).
    private func fireRecall(
        _ kit: GeniusLocusKit,
        _ handle: EstateHandle
    ) async throws {
        let request = GLKRecallRequest(
            frame: RecallFrame(filterChain: [.currentlyBelieve], limit: 50),
            mode: .locusOnly,
            scoring: .raw,
            limit: 50,
            fallback: .failClosed,
            origin: .external
        )
        _ = try await kit.recall(handle, request)
    }

    /// Build a governor optimised for deterministic tick() tests:
    /// - graphAnalyticsIntervalMs  = Int.max (never fires)
    /// - graphCentralityIntervalMs = Int.max (never fires)
    /// - preferenceIntervalMs      = Int.max (never fires)
    /// - topologyCadenceMs         = Int.max (never fires)
    /// - poolReduceCadenceMs       = Int.max (never fires)
    /// - poolDirectory/poolTableArtifactURL = nil (no pool I/O)
    /// - clock: injected — tick() uses its own `now` parameter, not the clock.
    private func makeGovernor(
        kit: GeniusLocusKit,
        handle: EstateHandle
    ) -> AutonomicGovernor {
        AutonomicGovernor(
            kit: kit,
            handle: handle,
            baseTickMs: 5_000,
            graphAnalyticsIntervalMs: Int.max,
            graphCentralityIntervalMs: Int.max,
            preferenceIntervalMs: Int.max,
            topologyCadenceMs: Int.max,
            poolReduceCadenceMs: Int.max,
            topologyHandler: nil,
            topologyFingerprintLoader: nil,
            topologyGate: nil,
            graphAnalyticsHandler: nil,
            poolDirectory: nil,
            poolTableArtifactURL: nil
        )
    }

    // MARK: - §1 POSITIVE: pending items + timer due → dreaming fires

    /// Verify that the governor fires the dreaming cycle when:
    ///   (a) the dreaming queue has pending items (external-origin recall fired), AND
    ///   (b) the timer interval (default 30 s) has elapsed.
    ///
    /// The pending-count gate sees count > 0, the timer gate is due →
    /// both pass → dreaming pump runs.
    @Test("§1: pending > 0 AND timer due → dreamingFired = true")
    func positiveQueuePendingAndTimerDueFires() async throws {
        let (kit, handle) = try await makeKit()

        // Capture two drawers so they co-surface on recall.
        try await capture(kit, handle, content: "REM-ALPHA Swift content A")
        try await capture(kit, handle, content: "REM-ALPHA Swift content B")

        // Fire one external-origin recall → one DreamingItem in the queue.
        try await fireRecall(kit, handle)

        // Confirm the pending count before ticking (non-claiming probe).
        let pending = await kit.dreamingQueuePendingCount(for: handle)
        #expect(pending != nil && pending! >= 1,
            "setup: expected ≥ 1 pending dreaming item, got \(pending as Any)")

        let gov = makeGovernor(kit: kit, handle: handle)

        // Tick at t = 31 s (past the 30 s default interval).
        // First-tick-fires behaviour: last_timer_fire is nil → timer due.
        // Pending gate: Some(1) > 0 → proceed.
        let t31 = Date(timeIntervalSince1970: 31)
        let report = await gov.tick(now: t31)

        #expect(report.dreamingFired,
            "§1 POSITIVE: pending > 0 + timer due → dreamingFired must be true")
    }

    // MARK: - §2 NEGATIVE: empty queue + timer due → dreaming does NOT fire

    /// Verify that the governor does NOT fire the dreaming cycle when the
    /// dreaming queue is empty (or not mounted), even when the timer is due.
    ///
    /// This is the anti-regression: without the §12.2 pending-count gate, the
    /// governor would call dreaming.pump() here and do wasteful work.
    /// The test MUST FAIL if the gate is removed.
    @Test("§2: empty queue (nil) + timer due → dreamingFired = false (anti-regression)")
    func negativeEmptyQueueTimerDueNoFire() async throws {
        let (kit, handle) = try await makeKit()

        // Capture drawers but do NOT fire any external-origin recall.
        // Dreaming queue is never mounted → pending returns nil.
        try await capture(kit, handle, content: "idle Swift content A")
        try await capture(kit, handle, content: "idle Swift content B")

        let pending = await kit.dreamingQueuePendingCount(for: handle)
        #expect(pending == nil,
            "setup: expected dreaming queue not mounted (nil), got \(pending as Any)")

        let gov = makeGovernor(kit: kit, handle: handle)

        // Tick at t = 31 s — timer is DUE (first-tick-fires: no prior fire).
        // Pending gate: nil → skip. dreaming must NOT fire.
        let t31 = Date(timeIntervalSince1970: 31)
        let report = await gov.tick(now: t31)

        #expect(!report.dreamingFired,
            "§2 NEGATIVE: empty queue (nil) + timer due → dreamingFired must be false")
    }

    // MARK: - §2b NEGATIVE: drained queue, second tick → dreaming does NOT fire

    /// Verify that after the queue is drained in the first cycle, a subsequent
    /// tick (timer due again) does NOT re-fire dreaming (queue is now empty).
    @Test("§2b: drained queue + timer due (second tick) → dreamingFired = false")
    func negativeDrainedQueueSecondTickNoFire() async throws {
        let (kit, handle) = try await makeKit()

        try await capture(kit, handle, content: "drain-once Swift A")
        try await capture(kit, handle, content: "drain-once Swift B")

        // Enqueue one item.
        try await fireRecall(kit, handle)

        let gov = makeGovernor(kit: kit, handle: handle)

        // First tick at t=31 s: timer due + pending=1 → fires + drains.
        let t31 = Date(timeIntervalSince1970: 31)
        let first = await gov.tick(now: t31)
        #expect(first.dreamingFired, "§2b setup: first tick (pending=1, timer due) must fire")

        // Verify queue is empty after drain.
        let afterDrain = await kit.dreamingQueuePendingCount(for: handle)
        #expect(afterDrain == 0,
            "§2b setup: after drain, pending must be 0, got \(afterDrain as Any)")

        // Second tick at t=62 s: timer due again (31 s > 30 s since last fire).
        // Queue empty (Some(0)) → pending gate skips.
        let t62 = Date(timeIntervalSince1970: 62)
        let second = await gov.tick(now: t62)

        #expect(!second.dreamingFired,
            "§2b NEGATIVE: drained queue + timer due on second tick → dreamingFired must be false")
    }

    // MARK: - §3 CADENCE: timer-mode cadence gate + NK-2 event-mode pumpOnEvent wiring

    /// Verify the cadence gate and NK-2 event-mode wiring together.
    ///
    /// After the first dreaming cycle fires at t=0, the SolverBandit updates the
    /// trigger mode based on the cycle's reward. If the bandit selects .timer mode,
    /// the timer gate holds at t=10 s (10 s < 30 s interval, timer not due →
    /// dreamingFired = false). If the bandit selects .event or .hybrid mode, the
    /// NK-2 pumpOnEvent path fires at t=10 s (pending > 0, threshold = 1 →
    /// dreamingFired = true) — this is CORRECT NK-2 behavior: event-mode estates
    /// receive near-realtime cycles as observations accumulate, independent of the
    /// timer cadence.
    ///
    /// The test reads `gov.dreamingTriggerMode()` after the first tick to determine
    /// which assertion applies. Both paths are validated.
    @Test("§3: cadence gate (timer mode) + NK-2 event-mode pumpOnEvent wiring")
    func cadenceTimerNotDueAndEventModeWiring() async throws {
        let (kit, handle) = try await makeKit()

        try await capture(kit, handle, content: "cadence Swift content A")
        try await capture(kit, handle, content: "cadence Swift content B")

        // Enqueue one item and fire the first tick to set the timer baseline.
        // The first tick fires dreaming (nil last fire → always fires) and drains.
        try await fireRecall(kit, handle)
        let gov = makeGovernor(kit: kit, handle: handle)

        let t0 = Date(timeIntervalSince1970: 0)
        let first = await gov.tick(now: t0)
        // The first tick fires dreaming (first-tick-fires) and drains the queue.
        // Timer baseline is now set at t=0. The bandit may update trigger mode.
        let _ = first.dreamingFired

        // Read the bandit-selected trigger mode AFTER the first cycle.
        let modeAfterFirstCycle = await gov.dreamingTriggerMode()

        // Enqueue a NEW item AFTER the first tick drained the queue.
        try await fireRecall(kit, handle)

        let pending = await kit.dreamingQueuePendingCount(for: handle)
        #expect(pending != nil && pending! >= 1,
            "§3 setup: expected ≥ 1 pending item after second enqueue, got \(pending as Any)")

        // Tick at t=10 s: timer baseline was set at t=0; elapsed = 10 s < 30 s.
        let t10 = Date(timeIntervalSince1970: 10)
        let report = await gov.tick(now: t10)

        switch modeAfterFirstCycle {
        case .timer:
            // Timer mode: timer NOT due (10 s < 30 s), pumpOnEvent returns nil →
            // dreamingFired must be false. This is the classic cadence gate.
            #expect(!report.dreamingFired,
                "§3 CADENCE (timer mode): pending > 0 but timer not due → dreamingFired must be false")
        case .event, .hybrid:
            // Event / hybrid mode: pumpOnEvent fires when pending ≥ threshold (1).
            // Timer NOT due only blocks the timer path, not the event path.
            // This is correct NK-2 behavior: near-realtime dreaming for event estates.
            #expect(report.dreamingFired,
                "§3 NK-2 (event/hybrid mode): pending > 0 → pumpOnEvent must fire → dreamingFired = true")
        }
    }
}

File: packages/kits/NeuronKit/rust/tests/dreaming_alpha_gate_tests.rs
//! REM-ALPHA pending-count gate tests.
//!
//! The §12.2 gate skips the dreaming cycle when the dreaming queue is empty
//! (or not yet mounted), EVEN when the timer interval is due. Only a non-empty
//! queue proceeds to the EstateDreamingReader snapshot + drain + decide path.
//!
//! Coverage:
//!   §1  POSITIVE:  dreaming queue has pending items AND timer due
//!                  → governor tick FIRES the dreaming cycle (dreaming_fired=true).
//!   §2  NEGATIVE:  dreaming queue EMPTY AND timer due
//!                  → governor tick does NOT fire (dreaming_fired=false).
//!                  This is the anti-regression: without the gate this tick
//!                  would run the reader + scan. The test MUST fail if the gate
//!                  is removed.
//!   §3  CADENCE:   dreaming queue has pending items BUT timer interval not
//!                  elapsed → governor tick does NOT fire (dreaming_fired=false).
//!
//! Infrastructure mirrors `dreaming_queue_drain_tests.rs`: live InMemoryDrawerStore
//! + EstateCoordinator, AutonomicGovernor::new_for_testing_with_pool (no pool
//! I/O, deterministic cadences).

use std::sync::{Arc, Mutex};
use std::time::{Duration, UNIX_EPOCH};

use genius_locus_kit::coordinator::EstateCoordinator;
use genius_locus_kit::handle::EstateHandle;
use genius_locus_kit::recall::{
    GLKRecallMode, GLKRecallRequest, GLKRecallScoring, RecallFallbackPolicy,
};
use locus_kit::drawer_operational::CaptureChannel;
use locus_kit::drawer_store::DrawerStore as LocusDrawerStore;
use locus_kit::drawer_store_inmemory::InMemoryDrawerStore;
use locus_kit::estate_types::{LatticeAnchor, OwnerCredentials};
use locus_kit::filter::{Filter, RecallFrame};
use locus_kit::frames::CaptureFrame;
use neuron_kit::autonomic_governor::AutonomicGovernor;

// ── Infrastructure ─────────────────────────────────────────────────────────────

/// Shared estate setup: one in-memory DrawerStore, opened into coordinator.
/// Returns the coordinator Arc (shared with governor), the handle, and the raw
/// Arc<dyn DrawerStore> the governor needs for sink construction.
fn make_estate() -> (Arc<Mutex<EstateCoordinator>>, EstateHandle, Arc<dyn LocusDrawerStore>) {
    let store: Arc<dyn LocusDrawerStore> =
        Arc::new(InMemoryDrawerStore::new(0, None).expect("InMemoryDrawerStore::new"));
    let mut coord = EstateCoordinator::new();
    let handle = coord
        .open(Arc::clone(&store), OwnerCredentials::new("alpha-gate-test"), 0, 100)
        .expect("coord.open");
    (Arc::new(Mutex::new(coord)), handle, store)
}

/// Capture a drawer into the estate via the coordinator.
fn capture_drawer(
    coord: &Arc<Mutex<EstateCoordinator>>,
    handle: &EstateHandle,
    content: &str,
    now: i64,
) {
    let frame = CaptureFrame::new(
        content,
        CaptureChannel::Typed,
        "test-room",
        LatticeAnchor::udc("000"),
        "alpha-gate-test",
        "no-embedding",
    );
    let c = coord.lock().unwrap();
    c.capture(handle, frame, now).expect("capture");
}

/// Fire one external-origin scored recall, which enqueues a DreamingItem
/// for the co-surfaced set (B-10a: enqueue fires ONLY on external origin).
fn fire_external_recall(
    coord: &Arc<Mutex<EstateCoordinator>>,
    handle: &EstateHandle,
    now: i64,
) {
    let req = GLKRecallRequest::new(RecallFrame::new(vec![Filter::Unconfirmed]))
        .with_mode(GLKRecallMode::LocusOnly)
        .with_scoring(GLKRecallScoring::Raw)
        .with_limit(50)
        .with_fallback(RecallFallbackPolicy::FailClosed)
        .external(); // B-10a boundary
    let c = coord.lock().unwrap();
    c.recall_scored(handle, req, now).expect("recall_scored");
}

/// Build an AutonomicGovernor configured for deterministic tests:
/// - topology_cadence_ms = u64::MAX (topology snapshot never fires — no sink)
/// - pool_reduce_cadence_ms = u64::MAX (no pool I/O in these tests)
/// - pool paths point to non-existent temp dirs (no filesystem touch)
fn make_governor(
    coord: Arc<Mutex<EstateCoordinator>>,
    handle: EstateHandle,
    store: Arc<dyn LocusDrawerStore>,
) -> AutonomicGovernor {
    // Disable pool reduce and topology duties so the tick is purely dreaming +
    // maintenance. Use paths that will never exist so a pool-reduce miss is a
    // no-op rather than an actual filesystem operation.
    let pool_dir = std::path::PathBuf::from("/tmp/alpha-gate-test-pool-NONEXISTENT");
    let pool_table = std::path::PathBuf::from("/tmp/alpha-gate-test-table-NONEXISTENT.bin");
    AutonomicGovernor::new_for_testing_with_pool(
        coord,
        handle,
        store,
        u64::MAX, // topology cadence: never
        None,     // no topology sink
        u64::MAX, // pool-reduce cadence: never
        pool_dir,
        pool_table,
    )
}

/// `SystemTime` at epoch + `secs` seconds. Used to feed deterministic `now`
/// into governor ticks without calling `SystemTime::now()`.
fn epoch_plus_secs(secs: u64) -> std::time::SystemTime {
    UNIX_EPOCH + Duration::from_secs(secs)
}

// ── §1 POSITIVE: pending > 0 AND timer due → dreaming_fired = true ─────────

/// Verify that the governor fires the dreaming cycle when:
///   (a) the dreaming queue has at least one pending item (≥1 external-origin recall fired),
///   (b) the timer interval (default 30 s) has elapsed.
///
/// This is the base case: recalls were made, the queue is loaded, the timer
/// is due — the cycle MUST run to drain the queue and process the co-recall set.
#[test]
fn t9_alpha_gate_positive_pending_and_timer_due_fires() {
    let (coord, handle, store) = make_estate();

    // Capture two drawers so they co-surface on external-origin recall.
    let t0: i64 = 1_000_000;
    capture_drawer(&coord, &handle, "REM-ALPHA content A", t0);
    capture_drawer(&coord, &handle, "REM-ALPHA content B", t0);

    // Fire one external-origin recall → one DreamingItem in the queue.
    // The pending count is now 1 (§12.2 gate will see Some(1) → proceed).
    fire_external_recall(&coord, &handle, t0);

    // Verify the item is in the queue before constructing the governor
    // (uses the test-seam accessor — not needed by production, just confirming
    // our test setup is correct).
    {
        let c = coord.lock().unwrap();
        let pending = c.dreaming_queue_pending_count_for_gate(&handle);
        assert_eq!(pending, Some(1), "setup: expected 1 pending dreaming item");
    }

    let mut gov = make_governor(Arc::clone(&coord), handle.clone(), Arc::clone(&store));

    // Tick at t=31 s — past the 30 s default dreaming interval.
    // Timer gate: due. Pending gate: Some(1) > 0. Both pass → cycle fires.
    let report = gov.tick(epoch_plus_secs(31));

    assert!(
        report.dreaming_fired,
        "§1 POSITIVE: pending items + timer due → dreaming_fired must be true"
    );
}

// ── §2 NEGATIVE: empty queue AND timer due → dreaming_fired = false ─────────

/// Verify that the governor does NOT fire the dreaming cycle when the dreaming
/// queue is empty — even when the timer interval has elapsed.
///
/// This is the anti-regression test: without the §12.2 pending-count gate,
/// the governor would build an EstateDreamingReader and run a scan here,
/// wasting work on an idle tick. The test MUST FAIL if the gate is removed.
///
/// Empty-queue scenario: no external-origin recall has fired for this estate,
/// so the dreaming queue is not mounted at all (None from the probe, which is
/// treated identically to Some(0) — skip).
#[test]
fn t9_alpha_gate_negative_empty_queue_timer_due_no_fire() {
    let (coord, handle, store) = make_estate();

    // Capture drawers but do NOT fire any external-origin recall.
    // The dreaming queue is never mounted → pending = None.
    let t0: i64 = 1_000_000;
    capture_drawer(&coord, &handle, "idle content A", t0);
    capture_drawer(&coord, &handle, "idle content B", t0);

    // Verify queue is not mounted (test-seam confirmation of setup).
    {
        let c = coord.lock().unwrap();
        let pending = c.dreaming_queue_pending_count_for_gate(&handle);
        assert_eq!(pending, None, "setup: expected queue not mounted (None)");
    }

    let mut gov = make_governor(Arc::clone(&coord), handle.clone(), Arc::clone(&store));

    // Tick at t=31 s — timer gate is DUE. But pending gate is None → skip.
    let report = gov.tick(epoch_plus_secs(31));

    assert!(
        !report.dreaming_fired,
        "§2 NEGATIVE: empty queue + timer due → dreaming_fired must be false \
         (the §12.2 gate must block the scan; this test fails if the gate is removed)"
    );
}

// ── §2b NEGATIVE: queue mounted but zero remaining → dreaming_fired = false ─

/// Verify the no-fire case when the queue WAS populated but was already fully
/// drained in a prior cycle. The governor sees Some(0) — also a skip.
///
/// After the first tick drains the queue, the second tick (timer due again)
/// must NOT re-fire the dreaming cycle (the drain-once semantics guarantee the
/// queue is empty).
#[test]
fn t9_alpha_gate_negative_drained_queue_second_tick_no_fire() {
    let (coord, handle, store) = make_estate();

    let t0: i64 = 1_000_000;
    capture_drawer(&coord, &handle, "drain-once content A", t0);
    capture_drawer(&coord, &handle, "drain-once content B", t0);

    // One external-origin recall → one DreamingItem.
    fire_external_recall(&coord, &handle, t0);

    let mut gov = make_governor(Arc::clone(&coord), handle.clone(), Arc::clone(&store));

    // First tick at t=31 s: timer due + queue non-empty → dreaming fires + drains.
    let first = gov.tick(epoch_plus_secs(31));
    assert!(
        first.dreaming_fired,
        "§2b setup: first tick (pending=1, timer due) must fire"
    );

    // After drain, the queue is empty. Second tick at t=62 s: timer due again,
    // but queue is empty (Some(0)) → pending gate blocks the cycle.
    let second = gov.tick(epoch_plus_secs(62));
    assert!(
        !second.dreaming_fired,
        "§2b NEGATIVE: drained queue + timer due on second tick → dreaming_fired must be false"
    );
}

// ── §3 CADENCE: timer-mode cadence gate + NK-2 event-mode pump_on_event wiring ──

/// Verify the cadence gate and NK-2 event-mode wiring together.
///
/// After the first dreaming cycle fires at t=0, the SolverBandit updates the
/// trigger mode based on the cycle's reward. If the bandit selects Timer mode,
/// the timer gate holds at t=10 s (10 s < 30 s interval, timer not due →
/// dreaming_fired = false). If the bandit selects Event or Hybrid mode, the
/// NK-2 pump_on_event path fires at t=10 s (pending > 0, threshold = 1 →
/// dreaming_fired = true) — this is CORRECT NK-2 behavior: event-mode estates
/// receive near-realtime cycles as observations accumulate, independent of the
/// timer cadence.
///
/// The test reads `gov.dreaming_trigger_mode()` after the first tick to determine
/// which assertion applies. Both paths are validated.
#[test]
fn t9_alpha_gate_cadence_timer_not_due_no_fire_despite_pending() {
    use neuron_kit::DreamingTriggerMode;

    let (coord, handle, store) = make_estate();

    let t0: i64 = 1_000_000;
    capture_drawer(&coord, &handle, "cadence content A", t0);
    capture_drawer(&coord, &handle, "cadence content B", t0);

    // Enqueue one dreaming item and fire at t=0 to SET the timer baseline.
    // The first tick always fires (None last_fire → true). After this tick
    // the dreaming queue is drained (empty) and the timer baseline is set at t=0.
    fire_external_recall(&coord, &handle, t0);

    let mut gov = make_governor(Arc::clone(&coord), handle.clone(), Arc::clone(&store));
    let first = gov.tick(epoch_plus_secs(0));
    // The first tick fires dreaming (first-tick-fires behaviour) and drains the queue.
    // The bandit may update trigger_mode after the cycle.
    let _ = first.dreaming_fired;

    // Read the bandit-selected trigger mode AFTER the first cycle.
    let mode_after_first_cycle = gov.dreaming_trigger_mode();

    // Now enqueue a SECOND item while the timer is NOT due yet.
    fire_external_recall(&coord, &handle, t0 + 5);

    {
        // Confirm the new item is pending.
        let c = coord.lock().unwrap();
        let pending = c.dreaming_queue_pending_count_for_gate(&handle);
        assert!(
            pending.is_some() && pending.unwrap() >= 1,
            "setup: expected ≥ 1 pending item after second enqueue, got {:?}", pending
        );
    }

    // Tick at t=10 s — BEFORE the 30 s interval.
    let report = gov.tick(epoch_plus_secs(10));

    match mode_after_first_cycle {
        DreamingTriggerMode::Timer => {
            // Timer mode: timer NOT due (10 s < 30 s), pump_on_event returns None →
            // dreaming_fired must be false. This is the classic cadence gate.
            assert!(
                !report.dreaming_fired,
                "§3 CADENCE (timer mode): pending > 0 but timer not due → dreaming_fired must be false"
            );
        }
        DreamingTriggerMode::Event | DreamingTriggerMode::Hybrid => {
            // Event / hybrid mode: pump_on_event fires when pending ≥ threshold (1).
            // Timer NOT due only blocks the timer path, not the event path.
            // This is correct NK-2 behavior: near-realtime dreaming for event estates.
            assert!(
                report.dreaming_fired,
                "§3 NK-2 (event/hybrid mode): pending > 0 → pump_on_event must fire → dreaming_fired = true"
            );
        }
    }
}
